File: packages/agent/src/posthog-api.ts
import packageJson from "../package.json" with { type: "json" };
import type {
  ArtifactType,
  PostHogAPIConfig,
  StoredEntry,
  Task,
  TaskRun,
  TaskRunArtifact,
} from "./types";
import {
  getGatewayInvalidatePlanCacheUrl,
  getGatewayUsageUrl,
  getLlmGatewayUrl,
} from "./utils/gateway";

export {
  getGatewayInvalidatePlanCacheUrl,
  getGatewayUsageUrl,
  getLlmGatewayUrl,
};

const DEFAULT_USER_AGENT = `posthog/agent.hog.dev; version: ${packageJson.version}`;

export interface TaskArtifactUploadPayload {
  name: string;
  type: ArtifactType;
  content: string;
  content_type?: string;
}

export type TaskRunUpdate = Partial<
  Pick<
    TaskRun,
    | "status"
    | "branch"
    | "stage"
    | "error_message"
    | "output"
    | "state"
    | "environment"
  >
> & {
  state_remove_keys?: string[];
};

export class PostHogAPIClient {
  private config: PostHogAPIConfig;

  constructor(config: PostHogAPIConfig) {
    this.config = config;
  }

  private get baseUrl(): string {
    const host = this.config.apiUrl.endsWith("/")
      ? this.config.apiUrl.slice(0, -1)
      : this.config.apiUrl;
    return host;
  }

  private isAuthFailure(status: number): boolean {
    return status === 401 || status === 403;
  }

  private async resolveApiKey(forceRefresh = false): Promise<string> {
    if (forceRefresh && this.config.refreshApiKey) {
      return this.config.refreshApiKey();
    }

    return this.config.getApiKey();
  }

  private async buildHeaders(
    options: RequestInit,
    forceRefresh = false,
  ): Promise<Headers> {
    const headers = new Headers(options.headers);
    headers.set(
      "Authorization",
      `Bearer ${await this.resolveApiKey(forceRefresh)}`,
    );
    headers.set("Content-Type", "application/json");
    headers.set("User-Agent", this.config.userAgent ?? DEFAULT_USER_AGENT);
    return headers;
  }

  private async performRequest(
    endpoint: string,
    options: RequestInit,
    forceRefresh = false,
  ): Promise<Response> {
    const url = `${this.baseUrl}${endpoint}`;

    return fetch(url, {
      ...options,
      headers: await this.buildHeaders(options, forceRefresh),
    });
  }

  private async performRequestWithRetry(
    endpoint: string,
    options: RequestInit = {},
  ): Promise<Response> {
    let response = await this.performRequest(endpoint, options);

    if (!response.ok && this.isAuthFailure(response.status)) {
      response = await this.performRequest(endpoint, options, true);
    }

    return response;
  }

  private async apiRequest<T>(
    endpoint: string,
    options: RequestInit = {},
  ): Promise<T> {
    const response = await this.performRequestWithRetry(endpoint, options);

    if (!response.ok) {
      let errorMessage: string;
      try {
        const errorResponse = await response.json();
        errorMessage = `Failed request: [${response.status}] ${JSON.stringify(errorResponse)}`;
      } catch {
        errorMessage = `Failed request: [${response.status}] ${response.statusText}`;
      }
      throw new Error(errorMessage);
    }

    return response.json();
  }

  private getTeamId(): number {
    return this.config.projectId;
  }

  async getApiKey(forceRefresh = false): Promise<string> {
    return this.resolveApiKey(forceRefresh);
  }

  getLlmGatewayUrl(): string {
    return getLlmGatewayUrl(this.baseUrl);
  }

  async getTask(taskId: string): Promise<Task> {
    const teamId = this.getTeamId();
    return this.apiRequest<Task>(`/api/projects/${teamId}/tasks/${taskId}/`);
  }

  async getTaskRun(taskId: string, runId: string): Promise<TaskRun> {
    const teamId = this.getTeamId();
    return this.apiRequest<TaskRun>(
      `/api/projects/${teamId}/tasks/${taskId}/runs/${runId}/`,
    );
  }

  async resumeRunInCloud(taskId: string, runId: string): Promise<TaskRun> {
    const teamId = this.getTeamId();
    return this.apiRequest<TaskRun>(
      `/api/projects/${teamId}/tasks/${taskId}/runs/${runId}/resume_in_cloud/`,
      { method: "POST" },
    );
  }

  async updateTaskRun(
    taskId: string,
    runId: string,
    payload: TaskRunUpdate,
  ): Promise<TaskRun> {
    const teamId = this.getTeamId();
    return this.apiRequest<TaskRun>(
      `/api/projects/${teamId}/tasks/${taskId}/runs/${runId}/`,
      {
        method: "PATCH",
        body: JSON.stringify(payload),
      },
    );
  }

  async setTaskRunOutput(
    taskId: string,
    runId: string,
    output: Record<string, unknown>,
  ): Promise<TaskRun> {
    return this.apiRequest(
      `/api/projects/${this.getTeamId()}/tasks/${taskId}/runs/${runId}/set_output/`,
      {
        method: "PATCH",
        body: JSON.stringify(output),
      },
    );
  }

  async appendTaskRunLog(
    taskId: string,
    runId: string,
    entries: StoredEntry[],
  ): Promise<TaskRun> {
    const teamId = this.getTeamId();
    return this.apiRequest<TaskRun>(
      `/api/projects/${teamId}/tasks/${taskId}/runs/${runId}/append_log/`,
      {
        method: "POST",
        body: JSON.stringify({ entries }),
      },
    );
  }

  async relayMessage(
    taskId: string,
    runId: string,
    text: string,
  ): Promise<void> {
    const teamId = this.getTeamId();
    await this.apiRequest<{ status: string }>(
      `/api/projects/${teamId}/tasks/${taskId}/runs/${runId}/relay_message/`,
      {
        method: "POST",
        body: JSON.stringify({ text }),
      },
    );
  }

  async uploadTaskArtifacts(
    taskId: string,
    runId: string,
    artifacts: TaskArtifactUploadPayload[],
  ): Promise<TaskRunArtifact[]> {
    if (!artifacts.length) {
      return [];
    }

    const teamId = this.getTeamId();
    const response = await this.apiRequest<{ artifacts: TaskRunArtifact[] }>(
      `/api/projects/${teamId}/tasks/${taskId}/runs/${runId}/artifacts/`,
      {
        method: "POST",
        body: JSON.stringify({ artifacts }),
      },
    );

    const manifest = response.artifacts ?? [];

    // The backend returns the full run artifact manifest after each upload.
    // Callers want the artifacts corresponding to this upload request only.
    return manifest.slice(-artifacts.length);
  }

  /** Signal reports the given task is associated with (via report task associations). */
  async getSignalReportIdsForTask(taskId: string): Promise<string[]> {
    const teamId = this.getTeamId();
    const response = await this.apiRequest<{ results?: { id: string }[] }>(
      `/api/projects/${teamId}/signals/reports/?task_id=${encodeURIComponent(taskId)}&limit=100`,
    );
    return (response.results ?? []).map((r) => r.id);
  }

  /**
   * Append a log artefact to a signal report, attributed to `taskId` via the
   * `X-PostHog-Task-Id` header (the server validates it against the token's team).
   */
  async createSignalReportArtefact(
    reportId: string,
    taskId: string,
    body: { artefact_type: string; content: Record<string, unknown> },
  ): Promise<void> {
    const teamId = this.getTeamId();
    await this.apiRequest(
      `/api/projects/${teamId}/signals/reports/${reportId}/artefacts/`,
      {
        method: "POST",
        body: JSON.stringify(body),
        headers: { "X-PostHog-Task-Id": taskId },
      },
    );
  }

  /**
   * Download artifact content by storage path
   * Streams the file through the PostHog backend so the sandbox does not need
   * direct access to object storage.
   */
  async downloadArtifact(
    taskId: string,
    runId: string,
    storagePath: string,
  ): Promise<ArrayBuffer | null> {
    const teamId = this.getTeamId();

    try {
      const response = await this.performRequestWithRetry(
        `/api/projects/${teamId}/tasks/${taskId}/runs/${runId}/artifacts/download/`,
        {
          method: "POST",
          body: JSON.stringify({ storage_path: storagePath }),
        },
      );
      if (!response.ok) {
        throw new Error(`Failed to download artifact: ${response.status}`);
      }
      return response.arrayBuffer();
    } catch {
      return null;
    }
  }

  /**
   * Fetch logs for a task run via the logs API endpoint
   * @param taskRun - The task run to fetch logs for
   * @returns Array of stored entries, or empty array if no logs available
   */
  async fetchTaskRunLogs(taskRun: TaskRun): Promise<StoredEntry[]> {
    const teamId = this.getTeamId();
    const endpoint = `/api/projects/${teamId}/tasks/${taskRun.task}/runs/${taskRun.id}/logs`;

    try {
      const response = await this.performRequestWithRetry(endpoint);

      if (!response.ok) {
        if (response.status === 404) {
          return [];
        }
        throw new Error(
          `Failed to fetch logs: ${response.status} ${response.statusText}`,
        );
      }

      const content = await response.text();

      if (!content.trim()) {
        return [];
      }

      // Parse newline-delimited JSON
      return content
        .trim()
        .split("\n")
        .map((line) => JSON.parse(line) as StoredEntry);
    } catch (error) {
      throw new Error(
        `Failed to fetch task run logs: ${error instanceof Error ? error.message : String(error)}`,
      );
    }
  }
}

File: rust/crates/agent-server/src/posthog_api.rs
//! PostHog Django API client.
//!
//! Port of the endpoints in `packages/agent/src/posthog-api.ts` that the
//! agent-server calls. Task/TaskRun are deserialized loosely — the server
//! only reads a handful of fields and must tolerate schema growth.

use serde::Deserialize;
use serde_json::{json, Value};

#[derive(Debug, Clone)]
pub struct PostHogApiClient {
    client: reqwest::Client,
    base_url: String,
    project_id: i64,
    api_key: String,
    user_agent: String,
}

#[derive(Debug, thiserror::Error)]
#[error("{0}")]
pub struct ApiError(pub String);

#[derive(Debug, Clone, Default, Deserialize)]
pub struct Task {
    pub id: Option<String>,
    pub title: Option<String>,
    pub description: Option<String>,
    #[serde(default)]
    pub internal: bool,
    pub origin_product: Option<String>,
    pub signal_report: Option<String>,
    pub json_schema: Option<Value>,
    pub created_by: Option<Value>,
}

#[derive(Debug, Clone, Default, Deserialize)]
pub struct TaskRun {
    pub id: Option<String>,
    pub status: Option<String>,
    pub state: Option<Value>,
    pub artifacts: Option<Value>,
    pub log_url: Option<String>,
}

impl TaskRun {
    pub fn state_string(&self, key: &str) -> Option<String> {
        self.state.as_ref()?.get(key)?.as_str().map(str::to_string)
    }

    pub fn state_bool(&self, key: &str) -> Option<bool> {
        self.state.as_ref()?.get(key)?.as_bool()
    }
}

impl PostHogApiClient {
    pub fn new(api_url: &str, project_id: i64, api_key: &str, version: &str) -> Self {
        Self {
            client: reqwest::Client::new(),
            base_url: api_url.trim_end_matches('/').to_string(),
            project_id,
            api_key: api_key.to_string(),
            user_agent: format!("posthog/cloud.hog.dev; version: {version}"),
        }
    }

    pub fn base_url(&self) -> &str {
        &self.base_url
    }

    pub fn project_id(&self) -> i64 {
        self.project_id
    }

    async fn request(
        &self,
        method: reqwest::Method,
        endpoint: &str,
        body: Option<Value>,
    ) -> Result<Value, ApiError> {
        let url = format!("{}{}", self.base_url, endpoint);
        let mut request = self
            .client
            .request(method, &url)
            .header("Authorization", format!("Bearer {}", self.api_key))
            .header("Content-Type", "application/json")
            .header("User-Agent", &self.user_agent);
        if let Some(body) = body {
            request = request.json(&body);
        }

        let response = request
            .send()
            .await
            .map_err(|err| ApiError(format!("Request to {endpoint} failed: {err}")))?;

        let status = response.status();
        if !status.is_success() {
            let text = response.text().await.unwrap_or_default();
            return Err(ApiError(format!(
                "Failed request: [{}] {}",
                status.as_u16(),
                text
            )));
        }

        response
            .json::<Value>()
            .await
            .map_err(|err| ApiError(format!("Invalid JSON from {endpoint}: {err}")))
    }

    pub async fn get_task(&self, task_id: &str) -> Result<Task, ApiError> {
        let value = self
            .request(
                reqwest::Method::GET,
                &format!("/api/projects/{}/tasks/{task_id}/", self.project_id),
                None,
            )
            .await?;
        serde_json::from_value(value).map_err(|err| ApiError(format!("Invalid Task: {err}")))
    }

    pub async fn get_task_run(&self, task_id: &str, run_id: &str) -> Result<TaskRun, ApiError> {
        let value = self
            .request(
                reqwest::Method::GET,
                &format!(
                    "/api/projects/{}/tasks/{task_id}/runs/{run_id}/",
                    self.project_id
                ),
                None,
            )
            .await?;
        serde_json::from_value(value).map_err(|err| ApiError(format!("Invalid TaskRun: {err}")))
    }

    pub async fn update_task_run(
        &self,
        task_id: &str,
        run_id: &str,
        payload: Value,
    ) -> Result<(), ApiError> {
        self.request(
            reqwest::Method::PATCH,
            &format!(
                "/api/projects/{}/tasks/{task_id}/runs/{run_id}/",
                self.project_id
            ),
            Some(payload),
        )
        .await?;
        Ok(())
    }

    pub async fn set_task_run_output(
        &self,
        task_id: &str,
        run_id: &str,
        output: Value,
    ) -> Result<(), ApiError> {
        self.request(
            reqwest::Method::PATCH,
            &format!(
                "/api/projects/{}/tasks/{task_id}/runs/{run_id}/set_output/",
                self.project_id
            ),
            Some(output),
        )
        .await?;
        Ok(())
    }

    /// `entries` are `StoredNotification` envelopes
    /// (`{type:"notification",timestamp,notification}`).
    pub async fn append_task_run_log(
        &self,
        task_id: &str,
        run_id: &str,
        entries: Vec<Value>,
    ) -> Result<(), ApiError> {
        self.request(
            reqwest::Method::POST,
            &format!(
                "/api/projects/{}/tasks/{task_id}/runs/{run_id}/append_log/",
                self.project_id
            ),
            Some(json!({ "entries": entries })),
        )
        .await?;
        Ok(())
    }

    pub async fn relay_message(
        &self,
        task_id: &str,
        run_id: &str,
        text: &str,
        text_parts: &[String],
    ) -> Result<(), ApiError> {
        // Send `text_parts` alongside the joined `text` so backends that
        // understand the new schema can pick just the post-last-tool-use
        // answer, while older backends still get the flat `text` field.
        let mut body = json!({ "text": text });
        if !text_parts.is_empty() {
            body["text_parts"] = json!(text_parts);
        }
        self.request(
            reqwest::Method::POST,
            &format!(
                "/api/projects/{}/tasks/{task_id}/runs/{run_id}/relay_message/",
                self.project_id
            ),
            Some(body),
        )
        .await?;
        Ok(())
    }
}
